### DISCUSSION TITLE
Mongoose on Nuttx RTOS - serving static files from FS

### DISCUSSION THREAD

**Discussion by koliqi** [cesanta/mongoose#1439] [Ask questions, get answers] [answered]
I want to try mongoose web server on Nuttx. When run program getting following error in screen:

NuttShell (NSH) NuttX-10.2.0
nsh> mongoose &
mongoose [6:100]
nsh> Setup network addresses
Mounting SPIFFS filesystem at target=/www with source=/dev/esp/partition/spiffs
 ram disk ret = 0  allocsize 32768
1970-01-01 00:00:09 3 mongoose.c:3356:mg_listen  1 accepting on 192.168.1.172:80 (port 80)
1970-01-01 00:00:09 3 mongoose.c:3258:accept_con 2 accepted 192.168.1.67:57858
1970-01-01 00:00:09 1 mongoose.c:1332:printdiren 2 stat(/www/jquery-3.5.1.min.js): 2
1970-01-01 00:00:09 1 mongoose.c:1332:printdiren 2 stat(/www/favicon.ico): 2
1970-01-01 00:00:09 1 mongoose.c:1332:printdiren 2 stat(/www/print.html): 2
1970-01-01 00:00:09 1 mongoose.c:1332:printdiren 2 stat(/www/table.html): 2
1970-01-01 00:00:09 1 mongoose.c:1332:printdiren 2 stat(/www/index.html): 2
1970-01-01 00:00:10 3 mongoose.c:3090:read_conn  2   00100000000011       0 -1/2048 err 79
1970-01-01 00:00:10 3 mongoose.c:3119:write_conn 2   00100000000111      56 -1 err 128

on web browser (192.168.1.67), instead of index.html file getting response  
<h1>Index of /</h1>
Name | Modified | Size
<address>Mongoose v.7.5</address>

OK, mongoose see files on directory /www but can't read and submit to client index.html, instead sending something else. Problem is, mongoose can't read properly files from spiffs (flat filesystem). 
Question: how to make possible to read files from spiffs, how to tell mongoose it is flat file system?

**Comment by cpq** [comment, marked as answer]
Mongoose uses "virtual" file system API to access files. By default, it implements a POSIX interface, and a "packed" interface. User can implement its own custom interface, using whatever filesystem available. The filesystem interface can be specified as a `struct mg_http_serve_opt :: fs` member. If it is left as NULL, then whatever available API is used (which is POSIX).

So, the answer to your question is: implement your own custom filesystem API, and specify it in the `struct mg_http_serve_opts`. You can take a look at the example at https://github.com/cesanta/mongoose/blob/2000752b2c636a5c5a57141aa80cd4c2588076b4/examples/esp32/main/main.c#L29-L32 - which uses a hack: a POSIX implementation gets reused, with only one function (stat) replaced. In your case, when you have a FS prefix, you need to implement all filesystem functions - see POSIX implementation at https://github.com/cesanta/mongoose/blob/2000752b2c636a5c5a57141aa80cd4c2588076b4/src/fs_posix.c and do a similar thing using SPIFFS.

Your source code should look like this:

```c
static int my_stat(const char *path, size_t *size, time_t *mtime) {
  ...
}
static void my_list(const char *path, void (*fn)(const char *, void *), void *userdata) {
  ...
}
static void *my_open(const char *path, int flags) {
  ...
}
static void my_close(void *fd) {
  ...
}
static size_t my_read(void *fd, void *buf, size_t len) {
  ...
}
static size_t my_write(void *fd, const void *buf, size_t len) {
  ...
}
static size_t my_seek(void *fd, size_t offset) {
  ...
}

static void cb(struct mg_connection *c, int ev, void *ev_data, void *fn_data) {
  if (ev == MG_EV_HTTP_MSG) {
    struct mg_http_message *hm = ev_data;
    struct mg_fs fs = {my_stat, my_list,  my_open, my_close, my_read, my_write, my_seek};
    struct mg_http_serve_opts opts = {.root_dir = "/www", .fs = &fs};
    mg_http_serve_dir(c, hm, &opts);
  }
}
```

**Comment by acassis** [comment]
Hi @koliqi and @cpq I can't believe I missed that issue post for almost 4 years.

Maybe it is better to use some NuttX more standard VFS like NXFFS or SmartFS.

BTW, if mongoose is working on NuttX we could add some recipes to add it to nuttx-apps and simplify the process to end users

**Reply by cpq** [reply]
Thanks @acassis 

Is it possible to retarget Nuttx's filesystem to a standard ISO C file API - fopen/fclose/fread/fwrite ?
That would be the simplest way, cause in that case nothing is required on the mongoose side.

**Reply by acassis** [reply]
@cpq yes, NuttX follows the POSIX standard, so all these functions are used by default to interact with the VFS (i.e. https://github.com/apache/nuttx/blob/master/fs/vfs/fs_open.c). I didn't understand exactly why he couldn't read the the setting from SPIFFS, maybe there are some issues in the SPIFFS integration on NuttX.

This is why I suggested testing using a native FS implemented from scratch on NuttX.
